### PR TITLE
add chaiscript

### PR DESCRIPTION
**Pull request by Garcia6l20**
Specify library name and version:  **chaiscript/6.1.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/chaiscript/all/test_package/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+project(test_package CXX)
+cmake_minimum_required(VERSION 2.8.12)
+
+set(CMAKE_VERBOSE_MAKEFILE ON)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup(TARGETS)
+
+find_package(chaiscript REQUIRED CONFIG)
+find_package(Threads)
```
Comment (uilianries) [inline recipes/chaiscript/all/test_package/CMakeLists.txt]: Why do you need it if you are exporting pthread in package_info?

File: recipes/chaiscript/all/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+# TEST_PACKAGE #################################################################
+
+if(NOT WIN32)
+    link_libraries(dl)
+endif()
+
+add_executable(${CMAKE_PROJECT_NAME} test_package.cpp)
+target_link_libraries(${CMAKE_PROJECT_NAME} CONAN_PKG::chaiscript Threads::Threads)
+set_property(TARGET ${CMAKE_PROJECT_NAME} PROPERTY CXX_STANDARD 17)
```
Comment (madebr) [inline recipes/chaiscript/all/test_package/CMakeLists.txt]: chaiscript requires a C++14 compiler, so I think 14 suffices.

File: recipes/chaiscript/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup(TARGETS)
+
+find_package(chaiscript REQUIRED CONFIG)
+find_package(Threads)
+
+# TEST_PACKAGE #################################################################
+
+if(NOT WIN32)
+    link_libraries(dl)
```
Comment (madebr) [inline recipes/chaiscript/all/test_package/CMakeLists.txt]: This must be removed and set in `package_info`.

File: recipes/chaiscript/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    name = "chaiscript"
+    homepage = "https://github.com/ChaiScript/ChaiScript"
+    description = "Embedded Scripting Language Designed for C++."
+    topics = ("conan", "embedded-scripting-language", "language")
+    url = "https://github.com/conan-io/conan-center-index"
+    license = "BSD-3-Clause"
+    exports_sources = ['CMakeLists.txt']
+    generators = 'cmake'
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"fPIC": [True, False], "dyn_load": [True, False], "use_std_make_shared": [True, False],
```
Comment (madebr) [inline recipes/chaiscript/all/conanfile.py]: No `shared` option?

File: recipes/chaiscript/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+class ChaiScriptConan(ConanFile):
+    name = "chaiscript"
+    homepage = "https://github.com/ChaiScript/ChaiScript"
+    description = "Embedded Scripting Language Designed for C++."
+    topics = ("conan", "embedded-scripting-language", "language")
+    url = "https://github.com/conan-io/conan-center-index"
+    license = "BSD-3-Clause"
+    exports_sources = ['CMakeLists.txt']
+    generators = 'cmake'
```
Comment (madebr) [inline recipes/chaiscript/all/conanfile.py]: *(style: use double quotes consistently)*

File: recipes/chaiscript/all/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.1.0)
+project(cmake_wrapper)
+
+include(conanbuildinfo.cmake)
+conan_basic_setup()
+
+if (WIN32 AND BUILD_SHARED_LIBS)
```
Comment (madebr) [inline recipes/chaiscript/all/CMakeLists.txt]: Unneeded, unless you add a `shared` option.

Comment (madebr) [reply recipes/chaiscript/all/CMakeLists.txt]: Also, `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` was introduced in cmake 3.4, so you would either need to increase the version on top, or make a new call here.

**Comment by uilianries** [comment]
Hi @Garcia6l20 ! Thanks for your contribution. Please, take a look on the comments.

File: recipes/chaiscript/all/test_package/CMakeLists.txt (hunk 1 above)
Comment (Garcia6l20) [reply recipes/chaiscript/all/test_package/CMakeLists.txt]: I'm still learning the conan's way...

File: recipes/chaiscript/all/test_package/CMakeLists.txt (hunk 2 above)
Comment (Garcia6l20) [reply recipes/chaiscript/all/test_package/CMakeLists.txt]: Oh yes, sorry I was looking at the develop branch

**Comment by Garcia6l20** [comment]
Thanks for reviewing and accepting me in EAP.

File: recipes/chaiscript/all/CMakeLists.txt (hunk 6 above)
Comment (Garcia6l20) [reply recipes/chaiscript/all/CMakeLists.txt]: Ok, I added the shared option and increased the minimum cmake version

File: recipes/chaiscript/all/conanfile.py (hunk 4 above)
Comment (Garcia6l20) [reply recipes/chaiscript/all/conanfile.py]: Added

File: recipes/chaiscript/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = "ChaiScript-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def _configure_cmake(self):
+        cmake = CMake(self)
```
Comment (uilianries) [inline recipes/chaiscript/all/conanfile.py]:
```suggestion
        cmake = CMake(self)
        cmake.definitions["BUILD_TESTING"] = False
```

Comment (Garcia6l20) [reply recipes/chaiscript/all/conanfile.py]: You're write, I'll make it header only

**Comment by uilianries** [comment]
You could add the option `header_only` which does not generate those modules.

**Comment by Garcia6l20** [comment]
Please notice the interpreter will also being packaged in non header_only package

**Comment by Garcia6l20** [comment]
Should I remove *header_only* option if *shared* is specified or raise a *ConanInvalidConfiguration* ?

**Comment by Garcia6l20** [comment]
Or just remove it, it's a non-sense in chaiscript case.
Header only + loadable modules

**Comment by Garcia6l20** [comment]
Unrelated question: are you interrested in [Boost.DI](https://github.com/boost-experimental/di), are experimental libraries ok to be available on conan-center ?

**Comment by uilianries** [comment]
Bincrafters already offers modular Boost, here in CCI we only have Boost package. Separated packages for Boost are really hard to be maintained/aligned, thus I don't think is a good idea. However, you can try it for Bincrafters.

**Pull request merged**